dask_kubernetes/core.py

```diff
@@ -453,46 +484,53 @@ def logs(self, pod=None):
-                await self._delete_pods(pending_to_delete)
-                n_to_delete = n_to_delete - len(pending_to_delete)
-                if n_to_delete <= 0:
-                    return
-
-            to_close = select_workers_to_close(self.scheduler, n_to_delete)
-            logger.debug("Closing workers: %s", to_close)
-            if len(to_close) < len(self.scheduler.workers):
-                # Close workers cleanly to migrate any temporary results to
-                # remaining workers.
-                @gen.coroutine
-                def f(to_close):
-                    yield self.scheduler.retire_workers(
-                        workers=to_close, remove=True, close_workers=True
-                    )
-                    yield offload(self.scale_down, to_close)
-
-                self.scheduler.loop.add_callback(f, to_close)
+        with log_errors():
+            if n >= len(self.scheduler.workers):
+                self.task_queue.put_nowait(self._scale_up)
+            else:
+                self.task_queue.put_nowait((self._scale_down_with_pending_check))
+
+
+    async def _scale_down_with_pending_check(self):
+        pods = await self._pods()
+        pods = await self._cleanup_terminated_pods(pods)
+        n = self._manual_scale_target
+        n_to_delete = len(pods) - n
+
+        # Before trying to close running workers, check if we can cancel
+        # pending pods (in case the kubernetes cluster was too full to
+        # provision those pods in the first place).
+        running_workers = list(self.scheduler.workers.keys())
+        running_ips = set(urlparse(worker).hostname for worker in running_workers)
+        pending_pods = [p for p in pods if p.status.pod_ip not in running_ips]
+        if pending_pods:
+            pending_to_delete = pending_pods[:n_to_delete]
+            logger.debug("Deleting pending pods: %s", pending_to_delete)
+            await self._delete_pods(pending_to_delete)
+            n_to_delete = n_to_delete - len(pending_to_delete)
+            if n_to_delete <= 0:
                 return
+        to_close = select_workers_to_close(self.scheduler, n_to_delete)
+        logger.debug("Closing workers: %s", to_close)
+        if len(to_close) < len(self.scheduler.workers):
+            # Close workers cleanly to migrate any temporary results to
+            # remaining workers.
+            @gen.coroutine
+            def f(to_close):
+                yield self.scheduler.retire_workers(
+                    workers=to_close, remove=True, close_workers=True
+                )
+                yield offload(self.scale_down, to_close)
+
+            self.scheduler.loop.add_callback(f, to_close)
+            return
 
-            # Terminate all pods without waiting for clean worker shutdown
-            self.scale_down(to_close)
+        # Terminate all pods without waiting for clean worker shutdown
+        return await self._scale_down(to_close)
```

Comment:
It seems like we want to wait for things to retire gracefully if we can, and then only shut things down hard if we weren't able to shut things down cleanly in a reasonable amount of time.

I would have expected us to just `await self.scheduler.retire_workers(...)` above rather than schedule it as a separate callback.

Reply:
Or maybe I should just say that there is value to waiting for the workers to transfer data before killing them.  Otherwise during adaptive workloads we end up repeating a lot of computation as our cluster scales itself down and loses results.